### PR TITLE
Token fetch Refresher optimization

### PR DESCRIPTION
**Pull request by Leah-Xia-Microsoft**
## Purpose

* This refactor is based on the suggestion from the calling SDK team
* for CommunicationTokenRefreshOptions, it's recommended not passing pass initialToken as nil
```
CommunicationTokenRefreshOptions(initialToken: nil, refreshProactively: true, tokenRefresher: tokenRefresher)
```
* The reason being that "the problem is create call agent is blocked unnecessarily if you do it before that you know whether contoso has properly working token refresh url or not instead of trying to debug why createcallagent is stuck"
* Also added activity indicator before launching the call

## Does this introduce a breaking change?

```
[x] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
* Get the code and set up with AAD.
* Start and Join a group/teams call

## What to Check
*Verify that we can start and join a group/teams call

### CODE REVIEW

File: AzureCalling/Controllers/InviteViewController.swift [hunk 1]
```diff
@@ -114,6 +115,11 @@ class InviteViewController: UIViewController {
 
     private func onContinueButtonTapped() {
         let callConfig = JoinCallConfig(joinId: groupCallId, displayName: displayName ?? "", callType: .groupCall)
-        callingContext.startCallComposite(callConfig)
+        busyOverlay.presentIn(view: view)
+        self.callingContext.startCallComposite(callConfig, completionHandler: { [weak self] in
+            DispatchQueue.main.async {
```
Comment (emlynmac) [inline AzureCalling/Controllers/InviteViewController.swift:120]: You should not need this - onContinueButtonTapped is called from the button action, which is from the main thread.

File: AzureCalling/Controllers/JoinCallViewController.swift [hunk 2]
```diff
@@ -166,7 +167,12 @@ class JoinCallViewController: UIViewController {
             return
         }
         let callConfig = JoinCallConfig(joinId: joinId, displayName: displayName ?? "", callType: joinCallType)
-        callingContext.startCallComposite(callConfig)
+        busyOverlay.presentIn(view: view)
+        self.callingContext.startCallComposite(callConfig, completionHandler: { [weak self] in
+            DispatchQueue.main.async {
```
Comment (emlynmac) [inline AzureCalling/Controllers/JoinCallViewController.swift:172]: Same here, navigateToCall is called from the main thread.
Generally, we don't want to async dispatch unless needed.

File: AzureCalling/External/Calling/CallingContext.swift [hunk 3]
```diff
@@ -27,51 +27,71 @@ final class CallingContext {
         self.tokenFetcher = tokenFetcher
     }
 
-    // MARK: Public API
+    // MARK: Private function
+    private func fetchInitialToken(completionHandler: @escaping (String?) -> Void) {
```
Comment (emlynmac) [inline AzureCalling/External/Calling/CallingContext.swift:31]: Can we make these `async` methods to follow the rest of the authentication handling?

File: AzureCalling/External/Calling/CallingContext.swift [hunk 4]
```diff
@@ -27,51 +27,71 @@ final class CallingContext {
...
+            } catch {
+                print("ERROR: It was not possible to create user credential.")
+                completionHandler(nil)
+            }
         }
-        return tokenCredential
     }
 
-    func startCallComposite(_ joinConfig: JoinCallConfig) {
+    func startCallComposite(_ joinConfig: JoinCallConfig, completionHandler: @escaping () -> Void) {
```
Comment (emlynmac) [inline AzureCalling/External/Calling/CallingContext.swift:59]: This would be greatly simplified to read wth async function.

**Review by emlynmac — COMMENTED** [review]
I think we can make the functions async to improve readability.
I wonder too, if we should move the silent token refresh down to the AAD handler

**Comment by Leah-Xia-Microsoft** [comment]
> I think we can make the functions async to improve readability. I wonder too, if we should move the silent token refresh down to the AAD handler

I think the tokenFetcher needed for the initialToken fetch is passed down to CallingContext from SceneDelegate. Seems convenient to have the initialToken in CallingContext other than AAD handler?

**Review by emlynmac — APPROVED** [review]
LGTM

**Pull request merged**